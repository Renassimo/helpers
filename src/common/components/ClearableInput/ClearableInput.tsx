import FormControl from '@mui/material/FormControl';
import Input, { InputProps } from '@mui/material/Input';
import InputAdornment from '@mui/material/InputAdornment';
import InputLabel from '@mui/material/InputLabel';

import IconButton from '@mui/material/IconButton';
import CloseIcon from '@mui/icons-material/Close';
import OpenInNewIcon from '@mui/icons-material/OpenInNew';
import RestartAltIcon from '@mui/icons-material/RestartAlt';

const ClearableInput = ({
  setValue,
  value,
  fullWidth = false,
  placeholder,
  label,
  type,
  disabled = false,
  originalValue,
  inputProps = {},
}: {
  setValue: (value: string | number | null) => void;
  value?: string | number | null;
  fullWidth?: boolean;
  placeholder?: string;
  label?: string;
  type?: string;
  disabled?: boolean;
  originalValue?: number | string | null;
  inputProps?: InputProps;
}) => {
  return (
    <FormControl fullWidth={fullWidth}>
      {label && <InputLabel variant="standard">{label}</InputLabel>}
      <Input
        {...inputProps}
        type={type}
        placeholder={placeholder}
        aria-label={inputProps['aria-label'] ?? label ?? placeholder}
        value={value ?? ''}
        fullWidth={fullWidth}
        disabled={disabled}
        onChange={(event) => setValue(event.target.value)}
        endAdornment={
          <InputAdornment position="end">
            {originalValue && originalValue !== value && (
              <IconButton
                aria-label={'Refresh'}
                onClick={() => setValue(originalValue ?? null)}
                edge="end"
              >
                {<RestartAltIcon />}
              </IconButton>
            )}
            {type === 'url' && value && (
              <IconButton
                aria-label={'url'}
                href={value as string}
                target="_blank"
                edge="end"
              >
                {<OpenInNewIcon />}
              </IconButton>
            )}
            {value && !disabled && (
              <IconButton
                aria-label="Clear"
                onClick={() => setValue('')}
                edge="end"
              >
                {<CloseIcon />}
              </IconButton>
            )}
          </InputAdornment>
        }
      />
    </FormControl>
  );
};

export default ClearableInput;
