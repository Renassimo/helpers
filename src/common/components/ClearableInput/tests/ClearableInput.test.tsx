import userEvent from '@testing-library/user-event';
import renderWithTheme from '@/common/tests/helpers/renderWithTheme';

import ClearableInput from '../ClearableInput';

describe('ClearableInput', () => {
  const setValue = jest.fn();
  const props = {
    setValue,
    value: 'value',
    fullWidth: true,
    // disabled: false,
    // originalValue: 'original value',
    inputProps: {
      'aria-label': 'Input',
    },
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders successfully', () => {
    // Arange
    // Act
    const { baseElement } = renderWithTheme(<ClearableInput {...props} />);
    // Assert
    expect(baseElement).toMatchSnapshot();
  });

  describe('when value is empty', () => {
    test('renders successfully', () => {
      // Arange
      // Act
      const { baseElement } = renderWithTheme(
        <ClearableInput {...props} value={null} label="Input label" />
      );
      // Assert
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('when type is url', () => {
    test('renders successfully', () => {
      // Arange
      // Act
      const { baseElement } = renderWithTheme(
        <ClearableInput {...props} type="url" placeholder="Placeholder" />
      );
      // Assert
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('when types', () => {
    test('calls setValue', async () => {
      // Arange
      const { getByLabelText } = renderWithTheme(<ClearableInput {...props} />);
      // Act
      await userEvent.type(getByLabelText('Input'), '1');
      // Assert
      expect(setValue).toBeCalledWith('value1');
    });
  });

  describe('when clears value', () => {
    test('calls setValue', async () => {
      // Arange
      const { getByLabelText } = renderWithTheme(<ClearableInput {...props} />);
      // Act
      await userEvent.click(getByLabelText('Clear'));
      // Assert
      expect(setValue).toBeCalledWith('');
    });
  });

  describe('when original value passed and clicks to refresh', () => {
    test('calls setValue', async () => {
      // Arange
      const originalValue = 'original value';
      const { getByLabelText } = renderWithTheme(
        <ClearableInput {...props} originalValue={originalValue} />
      );
      // Act
      await userEvent.click(getByLabelText('Refresh'));
      // Assert
      expect(setValue).toBeCalledWith(originalValue);
    });
  });
});
