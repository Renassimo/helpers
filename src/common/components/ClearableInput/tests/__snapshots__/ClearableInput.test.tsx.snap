// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClearableInput renders successfully 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
    >
      <div
        aria-label="Input"
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-953pxc-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd css-1x51dt5-MuiInputBase-input-MuiInput-input"
          type="text"
          value="value"
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        >
          <button
            aria-label="Clear"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ClearableInput when type is url renders successfully 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
    >
      <div
        aria-label="Input"
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-953pxc-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd css-1x51dt5-MuiInputBase-input-MuiInput-input"
          placeholder="Placeholder"
          type="url"
          value="value"
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        >
          <a
            aria-label="url"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
            href="value"
            tabindex="0"
            target="_blank"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="OpenInNewIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </a>
          <button
            aria-label="Clear"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ClearableInput when value is empty renders successfully 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
      >
        Input label
      </label>
      <div
        aria-label="Input"
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-953pxc-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd css-1x51dt5-MuiInputBase-input-MuiInput-input"
          type="text"
          value=""
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        />
      </div>
    </div>
  </div>
</body>
`;
